File: src/services/role-service.ts
import { NotFoundError } from "@/lib/errors";
import type { RoleRepository } from "@/repositories/role-repository";
import type { RoleDetail, RoleSummary } from "@/types/role";

export class RoleService {
  constructor(private readonly repository: RoleRepository) {}

  search(query: string): Promise<RoleSummary[]> {
    return this.repository.search(query);
  }

  async getById(id: string): Promise<RoleDetail> {
    const role = await this.repository.findById(id);
    if (!role) {
      throw new NotFoundError("Role");
    }

    const requirements = await this.repository.findRequirements([id]);
    return {
      ...role,
      requiredSkills: requirements.map((row) => row.requirement),
    };
  }
}

File: backend/src/main/java/com/talentgraph/service/RoleService.java
package com.talentgraph.service;

import com.talentgraph.exception.ResourceNotFoundException;
import com.talentgraph.model.RoleDetail;
import com.talentgraph.model.RoleRequirementRow;
import com.talentgraph.model.RoleSummary;
import com.talentgraph.model.SkillRequirement;
import com.talentgraph.repository.RoleRepository;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.stream.Collectors;

@Service
public class RoleService {
    private final RoleRepository repository;

    public RoleService(RoleRepository repository) {
        this.repository = repository;
    }

    public List<RoleSummary> search(String query) {
        return repository.search(query);
    }

    public RoleDetail getById(String id) {
        RoleSummary role = repository.findById(id);
        if (role == null) {
            throw new ResourceNotFoundException("Role not found");
        }

        List<RoleRequirementRow> requirements = repository.findRequirements(List.of(id));
        List<SkillRequirement> requiredSkills = requirements.stream()
            .map(RoleRequirementRow::requirement)
            .collect(Collectors.toList());

        return new RoleDetail(
            role.id(),
            role.slug(),
            role.name(),
            role.category(),
            role.seniority(),
            role.summary(),
            requiredSkills
        );
    }
}
